generative/3d_ldm/utils.py

```diff
@@ -0,0 +1,148 @@
+from monai.bundle import ConfigParser
+from monai.data import DataLoader
+from monai.transforms import (
+    AddChanneld,
+    CenterSpatialCropd,
+    Compose,
+    EnsureChannelFirstd,
+    EnsureTyped,
+    Lambdad,
+    LoadImaged,
+    Orientationd,
+    RandSpatialCropd,
+    ScaleIntensityRangePercentilesd,
+    Spacingd,
+)
+
+
+def setup_ddp(rank, world_size):
+    print(f"Running DDP diffusion example on rank {rank}/world_size {world_size}.")
+    print(f"Initing to IP {os.environ['MASTER_ADDR']}")
+    dist.init_process_group(
+        backend="nccl", init_method="env://", timeout=timedelta(seconds=36000), rank=rank, world_size=world_size
+    )  # gloo, nccl
+    dist.barrier()
+    device = torch.device(f"cuda:{rank}")
+    return dist, device
+
+
+def prepare_dataloader(
+    args,
+    batch_size,
+    patch_size,
+    randcrop=True,
+    rank=0,
+    world_size=1,
+    cache=1.0,
+    download=False,
+    size_divisible=16,
+    amp=False,
+):
+    ddp_bool = world_size > 1
+    channel = args.channel  # 0 = Flair, 1 = T1
+    assert channel in [0, 1, 2, 3], "Choose a valid channel"
+    if randcrop:
+        train_crop_transform = RandSpatialCropd(keys=["image"], roi_size=patch_size, random_size=False)
+        val_patch_size = [int(np.ceil(1.5 * p / size_divisible) * size_divisible) for p in patch_size]
+    else:
+        train_crop_transform = CenterSpatialCropd(keys=["image"], roi_size=patch_size)
+        val_patch_size = patch_size
+
+    if amp:
+        compute_dtype = torch.float16
+    else:
+        compute_dtype = torch.float32
+
+    train_transforms = Compose(
+        [
+            LoadImaged(keys=["image"]),
+            EnsureChannelFirstd(keys=["image"]),
+            Lambdad(keys="image", func=lambda x: x[channel, :, :, :]),
+            AddChanneld(keys=["image"]),
+            EnsureTyped(keys=["image"]),
+            Orientationd(keys=["image"], axcodes="RAS"),
+            Spacingd(keys=["image"], pixdim=args.spacing, mode=("bilinear")),
+            train_crop_transform,
+            ScaleIntensityRangePercentilesd(keys="image", lower=0, upper=99.5, b_min=0, b_max=1),
+            EnsureTyped(keys="image", dtype=compute_dtype),
+        ]
+    )
+    val_transforms = Compose(
+        [
+            LoadImaged(keys=["image"]),
+            EnsureChannelFirstd(keys=["image"]),
+            Lambdad(keys="image", func=lambda x: x[channel, :, :, :]),
+            AddChanneld(keys=["image"]),
+            EnsureTyped(keys=["image"]),
+            Orientationd(keys=["image"], axcodes="RAS"),
+            Spacingd(keys=["image"], pixdim=args.spacing, mode=("bilinear")),
+            CenterSpatialCropd(keys=["image"], roi_size=val_patch_size),
+            ScaleIntensityRangePercentilesd(keys="image", lower=0, upper=99.5, b_min=0, b_max=1),
+            EnsureTyped(keys="image", dtype=compute_dtype),
+        ]
+    )
+    train_ds = DecathlonDataset(
+        root_dir=args.data_base_dir,
```

Comment:
The `directory` seems ineffective because the root_dir is preset by `data_base_dir`.

Since the other tutorials presumes the data would be downloaded to `MONAI_DATA_DIRECTORY`, which usually is `/worksapce/data`, it would be good to re-use those downloaded datasets.

Reply:
This parameter is unused. I personally do not like the idea of hardcoding data path to main script... So shall I just remove `directory`?